[SuiteSparse_jll] Update to v7.4.0

giordano opened the pull request:
Usual memo to self for the update:

* change version number in `stdlib/SuiteSparse_jll/Project.toml`
* refresh checksums in `deps/checksums/suitesparse` with `make -f contrib/refresh_checksums.mk -j libsuitesparse`.  Note: this will also delete the files `deps/checksums/SuiteSparse-*.tar.gz/*`, but those refer to the stdlib which is the julia wrapper around the binary dependency, those files have to be restored manually, I don't know how to make the makefile behave nicely.  You will also have to remove the lines `SuiteSparse-*.tar.gz/md5/*` added to the file `deps/checksums/suitesparse`, they are wrong for the same reason
* update version numbers in `deps/libsuitesparse.version`.

Judging by the changes in https://github.com/JuliaPackaging/Yggdrasil/pull/7596 we don't need to touch the manual build system.

I have no idea whether SuiteSparse retains API/ABI compatibility within the same major version, I surely hope so because I won't have the strength to try and fix large compatibility errors.

giordano committed:
[SuiteSparse_jll] Upgrade to v7.4.0

giordano commented:
There is only one relevant test failure, of course on [32-bit Windows](https://buildkite.com/julialang/julia-master/builds/31305#018c7ed9-9a81-43ba-9a8c-5396ebbf073e/12580-13476):
```
Error in testset SparseArrays/cholmod:
Test Failed at C:\buildkite-agent\builds\win2k22-amdci6-4\julialang\julia-master\julia-b3081f8082\share\julia\stdlib\v1.11\SparseArrays\test\cholmod.jl:947
  Expression: (cholesky(A, perm = 1:n)).p == 1:n
   Evaluated: [12, 25, 5, 1, 2, 3, 4, 6, 7, 8  …  91, 92, 93, 94, 95, 96, 97, 98, 99, 100] == 1:100
```
I have no idea of what's the problem, so I'd appreciate any help here (also in the form of removing support for 32-bit Windows :upside_down_face:)

Everywhere else it's all good.  But we also need to merge https://github.com/JuliaSparse/SparseArrays.jl/pull/478 at the same time to clear the warnings about the version of cholmod:
```
      From worker 10:	┌ Warning: CHOLMOD version incompatibility
      From worker 10:	│
      From worker 10:	│ Julia was compiled with CHOLMOD version 4.0.4. It is
      From worker 10:	│ currently linked with version 5.0.0.
      From worker 10:	│ This might cause Julia to terminate when working with
      From worker 10:	│ sparse matrix factorizations, e.g. solving systems of
      From worker 10:	│ equations with \.
      From worker 10:	│
      From worker 10:	│ It is recommended that you use Julia with the same major
      From worker 10:	│ version of CHOLMOD as the one used during the build, or
      From worker 10:	│ download the generic binaries from www.julialang.org,
      From worker 10:	│ which ship with the correct versions of all dependencies.
      From worker 10:	└ @ SparseArrays.CHOLMOD /Users/julia/.julia/scratchspaces/a66863c6-20e8-4ff4-8a62-49f30b1f605e/agent-cache/default-honeycrisp-R17H3W25T9.0/build/default-honeycrisp-R17H3W25T9-0/julialang/julia-master/julia-b3081f8082/share/julia/stdlib/v1.11/SparseArrays/src/solvers/cholmod.jl:206
```

gbaraldi commented:
Is it adding some garbage to the start of the array?

giordano commented:
No, the number of elements is correct, but 12, 25, 5 for some reasons are at the front: https://buildkite.com/julialang/julia-master/builds/31334#018c82ce-35e2-43c1-a143-eebf240e7b15/12749-13075
```
F.p = [12, 25, 5, 1, 2, 3, 4, 6, 7, 8, 9, 10, 11, 13, 14, 15, 16, 17, 18, 19, 20, 21, 22, 23, 24, 26, 27, 28, 29, 30, 31, 32, 33, 34, 35, 36, 37, 38, 39, 40, 41, 42, 43, 44, 45, 46, 47, 48, 49, 50, 51, 52, 53, 54, 55, 56, 57, 58, 59, 60, 61, 62, 63, 64, 65, 66, 67, 68, 69, 70, 71, 72, 73, 74, 75, 76, 77, 78, 79, 80, 81, 82, 83, 84, 85, 86, 87, 88, 89, 90, 91, 92, 93, 94, 95, 96, 97, 98, 99, 100]
```
But this is wrong already at https://github.com/JuliaSparse/SparseArrays.jl/blob/8308232f12fe6d28c689e11b2e19b48a484136f6/src/solvers/cholmod.jl#L1320, and as far as I understand (but I may be wrong) the `symbolic` function is all written in Julia, no call into suitesparse yet: https://github.com/JuliaSparse/SparseArrays.jl/blob/8308232f12fe6d28c689e11b2e19b48a484136f6/src/solvers/cholmod.jl#L1258-L1278

ViralBShah commented:
cc @rayegun. The challenge here is of course how does one even test on 32-bit windows.

giordano commented:
I'm tempted by marking the test as broken on 32-bit Windows so that we can move on :smiling_face_with_tear:

ViralBShah commented:
I think it may actually be ok to do so.

rayegun commented:
😮‍💨, time to find the W32 VM again

giordano commented:
In case someone has a way to debug this on 32-bit Windows, I reduced the failure to
```julia
using LinearAlgebra, SparseArrays, SparseArrays.CHOLMOD

n = 100
A = sprand(n,n,5/n) |> t -> t't + I
SA = CHOLMOD.Sparse(A)
F = CHOLMOD.@cholmod_param postorder = false begin
    CHOLMOD.@cholmod_param nmethods = 1 begin
        CHOLMOD.analyze_p(SA, Int[p-1 for p in 1:n])
    end
end
F.p == 1:n # this should be true, but it's false
```
It can probably be reduced further, but I know absolutely nothing of the SparseArrays code, I think I reached the maximum I could spend on this.

If no one has ideas, I might just mark the test as broken only on that platforms so that we can upgrade the package.

ViralBShah commented:
@DrTimothyAldenDavis Would you by any chance have insight into what might be causing this failure on win32?

ViralBShah commented:
@DrTimothyAldenDavis Would you have any insight into what might be causing the win32 failure above?

DrTimothyAldenDavis commented:
No idea. 

I do try to maintain as much backward compatibility as possible between say a 4.0 and a 5.0 package. I would have to check the details to see what API/API differs. Is that the question?

I do test SuiteSparse in its CI with WIN32 on GitHub and it's passing. 

Can you try 7.4.0.betaX where X is 10 I think, but is changing daily?  I will release 7.4.0 stable soon.

ViralBShah commented:
That would suggest that there may be an issue in our wrappers.

DrTimothyAldenDavis commented:
Yes. It could be the cholmod 4 vs 5, above. 

Or it could be a Christmas bug. The first two numbers are 12,25 after all 😀

giordano commented:
There are tons of test failures with SuiteSparse v7.4.0....

giordano commented:
What's most worrying is that we're getting loads of
```
Error: no BLAS/LAPACK library loaded!
```
errors, which comes from libblastrampoline, but I have no clue of how this is related to upgrading to SuiteSparse 7.4.0

mmuetzel commented:
The build system of SuiteSparse has been completely overhauled for 7.4.0. Many of the configuration flags have changed their names. You'd probably need to adjust your build rules for that change here:
https://github.com/JuliaPackaging/Yggdrasil/blob/master/S/SuiteSparse/SuiteSparse%407/build_tarballs.jl

Maybe, you could also consider building with the root CMakeLists.txt now.

giordano commented:
> Many of the configuration flags have changed their names. You'd probably need to adjust your build rules for that change here:
[JuliaPackaging/Yggdrasil@master/S/SuiteSparse/SuiteSparse%407/build_tarballs.jl](https://github.com/JuliaPackaging/Yggdrasil/blob/master/S/SuiteSparse/SuiteSparse%407/build_tarballs.jl?rgh-link-date=2023-12-31T14%3A26%3A57Z)

Am I missing something or they were already updated by @imciner2 in 
https://github.com/JuliaPackaging/Yggdrasil/pull/7878?

DrTimothyAldenDavis commented:
This is wrong:


https://github.com/JuliaPackaging/Yggdrasil/blob/1210bc5fde582783d6dd87b5f36d4ad36fea42c8/S/SuiteSparse/SuiteSparse%407/build_tarballs.jl#L29-L37

if [[ ${nbits} == 64 ]]; then
    CMAKE_OPTIONS=(
        -DBLAS64_SUFFIX="_64"
        -DALLOW_64BIT_BLAS=YES
    )
else
    CMAKE_OPTIONS=(
        -DALLOW_64BIT_BLAS=NO
    )

This should refer to `SUITESPARSE_USE_64BIT_BLAS` not `ALLOW_64BIT_BLAS`

mmuetzel commented:
> > Many of the configuration flags have changed their names. You'd probably need to adjust your build rules for that change here:
> > [JuliaPackaging/Yggdrasil@master/S/SuiteSparse/SuiteSparse%407/build_tarballs.jl](https://github.com/JuliaPackaging/Yggdrasil/blob/master/S/SuiteSparse/SuiteSparse%407/build_tarballs.jl?rgh-link-date=2023-12-31T14%3A26%3A57Z)
> 
> Am I missing something or they were already updated by @imciner2 in [JuliaPackaging/Yggdrasil#7878](https://github.com/JuliaPackaging/Yggdrasil/pull/7878)?

I started to read that file from the top and came across `ALLOW_64BIT_BLAS`. That should be `SUITESPARSE_USE_64BIT_BLAS` now instead. Then, I was (probably erroneously) assuming that the build rules haven't been updated at all. Sorry if that was not the case.

mmuetzel commented:
Are the build logs accessible somewhere?

imciner2 commented:
I tried to get the ones that were obviously SuiteSparse build rules, but it almost looked like `ALLOW_64BIT_BLAS` was not SuiteSparse because we were also setting so many other CMake-internal BLAS rules to link against LBT.

Looking at a log here https://buildkite.com/julialang/yggdrasil/builds/7441#018cbca0-18bd-427d-9081-2e222ead3568, I guess it is complaining these aren't being used now:
* ALLOW_64BIT_BLAS
* BLAS_UNDERSCORE
* LAPACK_FOUND
* NPARTITION
* 
Which of these are SuiteSparse specific?

giordano commented:
> Are the build logs accessible somewhere?

https://buildkite.com/julialang/yggdrasil/builds/7441, or also in the release assets of https://github.com/JuliaBinaryWrappers/SuiteSparse_jll.jl/releases/tag/SuiteSparse-v7.4.0%2B0 (look at the files called `SuiteSparse-logs.*`, one for each platform).

mmuetzel commented:
> * ALLOW_64BIT_BLAS
> * BLAS_UNDERSCORE
> * LAPACK_FOUND
> * NPARTITION
> 
> Which of these are SuiteSparse specific?

Probably all of them are.
* `ALLOW_64BIT_BLAS`: Should be `SUITESPARSE_USE_64BIT_BLAS` now.
* `BLAS_UNDERSCORE`: Isn't a cmake flag but a preprocessor flag. Maybe, you intended to set the environment variable before calling CMake? E.g., `CPPFLAGS="-DBLAS_UNDERSCORE" cmake [...]`?
* `LAPACK_FOUND`: Not sure why that is set by the user. Maybe, it was needed in the past?
* `NPARTITION`: Should be `CHOLMOD_PARTITION` now. Note that the negated meaning of that flag has been removed.

I hope that helps.

mmuetzel commented:
Actually, I'm not sure if CMake picks up the value of the `CPPFLAGS` variable. Maybe, you'd need to set `CFLAGS` and/or `CXXFLAGS` instead...

DrTimothyAldenDavis commented:
> ALLOW_64BIT_BLAS
> BLAS_UNDERSCORE
> LAPACK_FOUND
> NPARTITION
> Which of these are SuiteSparse specific?

All but `LAPACK_FOUND` are SuiteSparse specific.

`BLAS_UNDERSCORE` is optional; how to call the BLAS is normally determined by the cmake scripts, which use the cmake FortranCInterface mechanism to find out how C can call Fortran.  Those can be overridden by setting `BLAS_NO_UNDERSCORE` or `BLAS_UNDERSCORE`, however.  

https://github.com/DrTimothyAldenDavis/SuiteSparse/blob/b72a742e9862547fe69cfa05de573ab33785607e/SuiteSparse_config/Config/SuiteSparse_config.h.in#L458-L494

DrTimothyAldenDavis commented:
Keep me posted if you come across any issues. Tag me in case I don't get a GitHub notification.

imciner2 commented:
The current issue with 7.4.0 is because the build system is not suffixing the BLAS/LAPACK symbols with the `_64` needed on the 64-bit BLAS platforms. @DrTimothyAldenDavis @mmuetzel I have opened https://github.com/DrTimothyAldenDavis/SuiteSparse/issues/668 in the SuiteSparse repo describing what I found.

DrTimothyAldenDavis commented:
The BLAS suffix issue should be fixed in the latest commits to the default (dev) branch on the SuiteSparse github repo.  Let me know if you'd like me to post a SuiteSparse 7.5.0.betaX pre-release with this change.

mmuetzel commented:
`SuiteSparse_long` changed it's type (more or less recently). It used to be always `long` in the past (apart from Windows). But now it can be `long` or `long long` depending on the platform and compiler. Maybe, that's related?

giordano commented:
From the fact the [build log of i686-linux-gnu](https://github.com/JuliaBinaryWrappers/SuiteSparse_jll.jl/releases/download/SuiteSparse-v7.4.0%2B2/SuiteSparse-logs.v7.4.0.i686-linux-gnu.tar.gz) doesn't mention "long" anywhere I presume the value of `SuiteSparse_long` isn't explicitly printed anywhere?  But I'd hope `SuiteSparse_long` doesn't magically become `long long` when targeting a 32-bit system (unless the decision is based on the host system, rather then the target one).  How do we check what SuiteSparse decided to do?

mmuetzel commented:
> But I'd hope `SuiteSparse_long` doesn't magically become `long long` when targeting a 32-bit system (unless the decision is based on the host system, rather then the target one). How do we check what SuiteSparse decided to do?

Unfortunately it does.
I had repeated discussions about that with @DrTimothyAldenDavis. But that's how he'd like to handle it now... 😞

I don't know of a good way to check that. Maybe, check `sizeof(SuiteSparse_long)` (if that's possible for you).

imciner2 commented:
Is `SuiteSparse_long` used in the code anymore? A search through the code doesn't show any uses of it, just the define inside `SuiteSparse_config.h` for it. All the changelogs mention switching to `int64_t` for all its old uses, so I guess it is unconditionally `int64_t` now? Did that change affect the API at all?

imciner2 commented:
So, I spent some time on this today (well... more like almost all afternoon), and here is a reproducer that triggers the error message on i686-linux-gnu:

```julia
using Test
using SparseArrays.CHOLMOD
using SparseArrays.CHOLMOD: getcommon
using LinearAlgebra:
    I, cholesky, cholesky!, det, diag, eigmax, ishermitian, isposdef, issuccess,
    issymmetric, ldlt, ldlt!, logdet, norm, opnorm, Diagonal, Hermitian, Symmetric,
    PosDefException, ZeroPivotException, RowMaximum
using SparseArrays
using SparseArrays: getcolptr
using SparseArrays.LibSuiteSparse
using SparseArrays.LibSuiteSparse: cholmod_l_allocate_sparse, cholmod_allocate_sparse

using InteractiveUtils

Ti = Int32
Tv = Float64

m = 27
n = 51
colptr0 = Ti[0,1,2,3,4,5,6,7,8,9,10,11,12,13,14,15,16,17,18,19,
23,25,27,29,33,37,41,45,47,49,51,53,55,57,59,63,65,67,69,71,75,79,83,87,89,
91,93,95,97,99,101,102]
rowval0 = Ti[2,3,6,7,8,9,12,13,16,17,18,19,20,21,22,23,24,25,26,
0,1,2,23,0,3,0,21,1,25,4,5,6,24,4,5,7,24,4,5,8,24,4,5,9,24,6,20,7,20,8,20,9,
20,3,4,4,22,5,26,10,11,12,21,10,13,10,23,10,20,11,25,14,15,16,22,14,15,17,
22,14,15,18,22,14,15,19,22,16,20,17,20,18,20,19,20,13,15,15,24,14,26,15]
nzval = Tv[1.0,1.0,1.0,1.0,1.0,1.0,1.0,1.0,1.0,1.0,1.0,1.0,1.0,1.0,1.0,1.0,1.0,1.0,
1.0,-1.0,-1.06,1.0,0.301,1.0,-1.0,1.0,-1.0,1.0,1.0,-1.0,-1.06,1.0,0.301,
-1.0,-1.06,1.0,0.313,-1.0,-0.96,1.0,0.313,-1.0,-0.86,1.0,0.326,-1.0,2.364,
-1.0,2.386,-1.0,2.408,-1.0,2.429,1.4,1.0,1.0,-1.0,1.0,1.0,-1.0,-0.43,1.0,
0.109,1.0,-1.0,1.0,-1.0,1.0,-1.0,1.0,1.0,-0.43,1.0,1.0,0.109,-0.43,1.0,1.0,
0.108,-0.39,1.0,1.0,0.108,-0.37,1.0,1.0,0.107,-1.0,2.191,-1.0,2.219,-1.0,
2.249,-1.0,2.279,1.4,-1.0,1.0,-1.0,1.0,1.0,1.0]
stype = 0

iss = true
for i = 2:length(colptr0)
    if !issorted(view(rowval0, colptr0[i - 1] + 1:colptr0[i]))
        iss = false
        break
    end
end


CHOLMOD.cholmod_allocate_sparse(m, n, colptr0[n + 1], iss, true, stype, CHOLMOD.xtyp(Tv), CHOLMOD.getcommon(Ti))
```

In here, the last call to the `cholmod_allocate_sparse` is failing, but the call to `getcommon` is not. The `cholmod_allocate_sparse` function is just a plain wrapper over the ccall into `libsuitesparse`.

I am not sure where to go with the debugging from here though. I stared at the CHOLMOD memory allocation code for a while, and nothing jumped out at me as being problematic on i686 architectures (e.g. using `int32_t` instead of `int64_t` for the indexing).

Anyone have ideas for things to try after this?

ViralBShah commented:
Perhaps the SuiteSparse wrappers need to be regenerated for 7.4?

giordano commented:
> Perhaps the SuiteSparse wrappers need to be regenerated for 7.4? I think those need refreshing anyways with every version bump.

How does one do that?

ViralBShah commented:
> > Perhaps the SuiteSparse wrappers need to be regenerated for 7.4? I think those need refreshing anyways with every version bump.
> 
> How does one do that?

Instructions on running the wrapper generator: https://github.com/JuliaSparse/SparseArrays.jl/tree/main/gen

ViralBShah commented:
I'm unable to regenerate the wrappers - it seems that Julia 1.10 does not have the SuiteSparse header files.

rayegun commented:
> I'm unable to regenerate the wrappers - it seems that Julia 1.10 does not have the SuiteSparse header files. Perhaps this should be discussed in a new issue on SparseArrays.jl

Fairly certain you need this PR as the Julia version you're using.

rayegun commented:
Who should we talk to about consolidating most of the metadata into SparseArrays.jl? Is that feasible eventually since the stdlibs are slowly moving out? JLL hashes etc should be based on the version of the JLL specified in SparseArrays.jl, not this current circular system.

ViralBShah commented:
IIRC, @KristofferC said we can never move SparseArrays.jl out of stdlib because it would be breaking. He would perhaps also be the right person to help resolve the metadata issue.

It has become impossible to maintain the SuiteSparse bindings - I have to re-learn the whole build system every few months.

rayegun commented:
Specifically @KristofferC we need the ability to version SuiteSparse_jll in the SparseArrays.jl repository, not in this repository. Any changes to SparseArrays.jl that touch the solvers need to essentially break tests here, before being fixed in SparseArrays.jl

imciner2 commented:
The SuiteSparse headers are now inside `include/suitesparse` in the jll, so yoi probably need to retarget the generator to the new location.

rayegun commented:
Hmmm this PR is not cooperating for me, I get errors with LibGit JLL. Can I download the JLL to a folder directly somehow? Or I can just use my local SuiteSparse

rayegun commented:
https://github.com/JuliaSparse/SparseArrays.jl/pull/480 Now we play chicken. Do we merge this first, or that :P?

rayegun commented:
The errors we see here are almost certainly since the `cholmod_common` struct changed its layout.

giordano committed:
Update wrappers in `SparseArrays.jl`

Update information in `stdlib/SparseArrays.version` and then run the commands

```
git rm -rf deps/checksums/SparseArrays-*
make -C stdlib -j checksum-SparseArrays
```

DrTimothyAldenDavis commented:
I've tried to keep the cholmod_common struct identical, or at least backward compatible.  I had even left in padding for future growth, which I've used since then.  It has grown in size at the end as well.

I extracted the definitions from CHOLMOD 3.0.14 in SuiteSparse 5.13.0, and the latest one in CHOLMOD 5.1.1 in SuiteSparse 7.5.0.beta, and then stripped the comments for a shorter diff:

[5.13.0.h.txt](https://github.com/JuliaLang/julia/files/13814421/5.13.0.h.txt)
[7.5.0.h.txt](https://github.com/JuliaLang/julia/files/13814422/7.5.0.h.txt)

The diff is very small, but I see one thing that I had meant to keep unchanged but somehow didn't.    The other_2 [4] array has changed from size_t to double.  Sorry about that.  That change only impacts a system where sizeof(size_t) is not sizeof(double), or a 32-bit hardware in other words.

The other change is to remove `SuiteSparse_long`.  That variable was always intended to be an `int64_t`, but it was before `int64_t` existed and it turned out that Windows can make it 32-bit.  That's why I've deprecated `SuiteSparse_long`; it was never supposed to be different sizes on different platforms.

Either of these changes could cause a 32-bit OS to see a different layout of the cholmod_common struct.

giordano commented:
Great news: after updating the wrappers using the SuiteSparse header files (PR https://github.com/JuliaSparse/SparseArrays.jl/pull/480 by @rayegun), all SparseArrays tests are passing in this PR, for all platforms, including 32-bit ones (and including the permutation test). The failures reported above were likely due to some changes in SuiteSparse between v7.2.1 and v7.3.0/v7.4.0 which broke ABI compatibility on 32-bit systems. Also @imciner2 identified the change to the type of other_2 in cholmod_common as a potential problem on 32-bit platforms, so that may be the culprit, and what forced us to update the wrappers.

Tomorrow I'll update https://github.com/JuliaSparse/SparseArrays.jl/pull/478 to include the changes in https://github.com/JuliaSparse/SparseArrays.jl/pull/480 and keep everything in one place, but this is now looking good for us, thanks everybody for the help!

DrTimothyAldenDavis commented:
> It has become impossible to maintain the SuiteSparse bindings - I have to re-learn the whole build system every few months.

Sorry for the headaches.  We're getting feedback from spack, PETsc, linux distros, msys, octave, ... and of course julia, and the feedback doesn't come all at once.  We're trying to make the build system work for everybody, but it means that it's an iterative process.

DrTimothyAldenDavis commented:
> The failures reported above were likely due to some changes in SuiteSparse between v7.2.1 and v7.3.0/v7.4.0 which broke ABI compatibility on 32-bit systems.

Whenever I change a package, I do keep to the semantic versioning process:  CHOLMOD has had a few breaking changes in the past, for example.  SuiteSparse itself doesn't track each version change of the packages inside it.  I used to bump the major SO of SuiteSparse itself when any of its component packages had a major SO bump, but that caused problems with linux distro maintainers so I don't do that anymore.

So here's how CHOLMOD has bumped in is major SO numbers, recently:

```
SuiteSparse     CHOLMOD
5.13.0          3.0.14      many cholmod*h include files
6.0.0           4.0.0       single cholmod.h include file, SuiteSparse_long deprecated
7.0.0           4.0.3
7.2.1           4.2.1
7.3.0           5.0.0       added single precision support (xworksize -> xworkbytes)
7.4.0           5.1.0
7.5.0           5.1.1
```

and you're note that the bump from CHOLMOD 4.x.x to 5.x.x did not cause SuiteSparse itself to bump from 7.2.1 to 8.0.0.

I may have broken the ABI upward compatibility by changing the `size_t other_2` to `double other_2`, however.  That was inadvertent.  That happened when going from SuiteSparse 7.2.1 to 7.3.0 and CHOLMOD bumped from 4.2.1 to 5.0.0 so at least that is "safe" in the sense that a breaking change to the ABI did get reflected in the major SO number.

DrTimothyAldenDavis commented:
For example, the change from `ALLOW_64BIT_BLAS` to `SUITESPARSE_USE_64BIT_BLAS` was from linux distro feedback, or spack (I can't recall exactly), so that all user-facing cmake variables are prefixed with `SUITESPARSE_` or `Package_` (like `UMFPACK_`, etc).

imciner2 commented:
> Sorry for the headaches.

No worries @DrTimothyAldenDavis. Thanks for the work you are doing to coalesce all these various requests and update the packages, the actual build system changes recently have actually led to a simplification of the build script we use in Yggdrasil, so that is appreciated.

One thing that might be useful going forward for doing the updates to these wrappers is to have a more detailed changelog for the APIs of the various components. For instance, when I started looking at the CHOLMOD 4->5 version bump to see if that was the cause of the errors, there wasn't really a good list of what changes were made to the API, so I ended up having to read through the header file and sometimes do a comparison between the old header and the new header. Some of the changes (like changing the naming/meaning of `xdtype` from `xtype` in `cholmod_allocate_sparse` were more subtle, and weren't obvious at first read through. Having a changelog for the API would make the process of updating easier and flag those types of subtle changes.

ViralBShah commented:
> > It has become impossible to maintain the SuiteSparse bindings - I have to re-learn the whole build system every few months.
> 
> Sorry for the headaches. We're getting feedback from spack, PETsc, linux distros, msys, octave, ... and of course julia, and the feedback doesn't come all at once. We're trying to make the build system work for everybody, but it means that it's an iterative process.

@DrTimothyAldenDavis Apologies - My comment was not meant for you, but for the way we integrate SuiteSparse in Julia itself and update it.

DrTimothyAldenDavis commented:
> One thing that might be useful going forward for doing the updates to these wrappers is to have a more detailed changelog for the APIs of the various components.

I can work on that.

giordano commented:
I updated this PR to use latest changes in https://github.com/JuliaSparse/SparseArrays.jl/pull/478. That PR now includes better way to update `SuiteSparse_jll` to regenerate our bindings, and also improved documentation to explain how to upgrade the version of the JLL package, this should address Viral's concerns about maintenance of our bindings.

If tests are green, as I hope since there should be no functional changes compared to this [fully green run](https://buildkite.com/julialang/julia-master/builds/31712) (excluding to may-fail jobs), we should be good to go.

giordano committed:
Update SuiteSparse build recipe for when `USE_BINARYBUILDER_LIBSUITESPARSE=0`

rayegun commented:
I am not sure off the top of my head how to regenerate the hashes, if no one does it before ~10AM tomorrow I will givve it a shot.

ViralBShah commented:
The checksum refreshing script is in `contrib/refresh_checksums.mk`. I can probably get to it tomorrow.

giordano commented:
> I am not sure off the top of my head how to regenerate the hashes

Update information in `stdlib/SparseArrays.version` and then run the commands

```
git rm -rf deps/checksums/SparseArrays-*
make -C stdlib -j checksum-SparseArrays
```

> if no one does it before ~10AM tomorrow I will givve it a shot.

I just did it.  This should be good to go now!

ViralBShah commented:
Does anyone know if the BLAS patch we are applying upstreamed to SuiteSparse?

DrTimothyAldenDavis commented:
> Does anyone know if the BLAS patch we are applying upstreamed to SuiteSparse?

I could take a look, and consider it for SuiteSparse 7.5.0.  Where is it in SuiteSparse_jll?

giordano commented:
It's a backport of https://github.com/DrTimothyAldenDavis/SuiteSparse/pull/671, it's already in SuiteSparse.

KristofferC commented:
> IIRC, @KristofferC said we can never move SparseArrays.jl out of stdlib because it would be breaking.

If I did I was just reiterating what triage (in its great wisdom) decided https://github.com/JuliaLang/julia/issues/48161#issuecomment-1414319635. So according to that we are indeed never allowed to remove stdlibs from Julia. We can of course just make them upgradable and ignore the bundled ones.
